Create Exports

bfcoder opened the pull request.

bfcoder committed:
Create all exports
Try 5 times
Log the good and the bad.
Delete existing exports before creating a new export.
Download exports
Move boot to lib workers
Update readme
Remove useless spec

bfcoder committed:
Move rake task to be included with gem

bfcoder committed:
Rename variables more appropriately

bfcoder committed:
Pass in the exports page to wait for export link

bfcoder committed:
Add specs

bfcoder committed:
Use appropriate Mutex for different ruby versions

bfcoder committed:
Add rdoc comments

jbasdf commented on bin/monitor:
```diff
@@ -1,7 +1,8 @@
 require "rack"
 
-Rack::Server.start config: "lib/scry/sidekiq/config.ru"
+Rack::Server.start config: File.join(LIB_DIR, "scry/sidekiq/config.ru")
```
Consider joining each piece of the path to be compatible with windows.

jbasdf commented on lib/scry/course.rb:
```diff
@@ -0,0 +1,163 @@
+        export_links = exports_page.links_with(
+          text: /Export Package/,
+        )
+        if export_links.any?
```
Let's add an else condition where we log the courses that don't have an export button.

jamis commented on lib/scry/course.rb:
```diff
@@ -0,0 +1,163 @@
+    def _process_export_form(export_page)
+      export_page.form_with(name: "selectCourse") do |export_form|
+        export_form.radiobuttons[1].check
+        export_form.radiobuttons[3].check
```
In the interest of avoiding magic numbers, and maybe making the code a bit easier to understand, maybe create two constants for `1` and `3` here, which describe what the radio buttons are that are being checked.

jamis commented on lib/scry/course.rb:
```diff
@@ -0,0 +1,163 @@
+        exports = exports_page.links_with(
+          text: "View Basic Log",
+        )
+        puts "waiting for link"
```
I noticed when you were running this with multiple exports in parallel, the "waiting for link" line got a bit spammy and wasn't super informative. What if you indicated which job was doing the waiting (maybe as simply as `Process.pid`)  and an indicator of how long the job has been waiting -- either as seconds since `_wait_for_export` was invoked, or as a count of how many times the loop has run?

jamis commented on lib/scry/scraper.rb:
```diff
@@ -1,5 +1,42 @@
+    agent.get(url) do |home_page|
+      index_page = home_page.form_with(name: "login") do |form|
+        form.fields[0].value = login
+        form.fields[1].value = passwd
```
As mentioned before, a constant for the `0` and `1` would avoid magic numbers and help clear up the intention here.

bfcoder committed:
Update readme

bfcoder committed:
Remove log files from gitignore

bfcoder committed:
Better file join

bfcoder committed:
2 spaces instead of 4 ugly spaces

bfcoder committed:
No magic numbers, use form specific elements
Display more info when exporting files to the terminal screen
Write to the log file if a course doesn’t have an export button
Improve rdoc for course.rb

bfcoder committed:
Use sidekiq-limit_fetch gem to limit queue workers

bfcoder committed:
space

bfcoder committed:
only get course_id once

bfcoder committed:
Add a two hour timeout to waiting for the exports

bfcoder committed:
Add courseId form to exports_page fixture

bfcoder commented on README.md:
````diff
@@ -20,30 +25,61 @@ Or install it yourself as:
+    scry_export_generator: 5
+    scry_export_downloader: 15
+```
+_note: limits is available through the sidekiq-limit_fetch gem_
````
Make a hyperlink

bfcoder commented on lib/scry/scraper.rb:
```diff
@@ -1,5 +1,42 @@
+        page: index_page,
+        text: /Open Bb Course List/,
+      )
+      course_links = courses_page.links_with(href: /type=Course/)
```
Add comment about assuming the courses are not paginated

bfcoder committed:
properly ignore blackboard_exports folder

bfcoder committed:
Add a log writer worker that is limited to one worker
- This will ensure that each log file can only be opened by one process.

bfcoder committed:
hyperlink the sidekiq-limit_fetch reference in the readme

bfcoder committed:
Add comment about unknown pagination for courses

bfcoder committed:
Add default/configurable log file names

bfcoder committed:
Only scrape courses that aren’t already downloaded.

bfcoder committed:
Move url, login, passwd  into scry.yml config

bfcoder committed:
fix specs for scraper now that params are not passed in

bfcoder committed:
Update course to use default config dir
include scry in workers to get default configs

bfcoder committed:
update readme

bfcoder committed:
require scry in course

bfcoder committed:
Verify scry.yml exists

bfcoder committed:
Use a method to get courses_downloaded
stub that method

bfcoder committed:
require scry in scraper

bfcoder committed:
Still scrape if no log file exists

bfcoder committed:
Update comments

bfcoder committed:
Check page is a Mechanize::Page before attempting to validate it.

srphillips commented:
This is sphillips and I approve this merge request

srphillips reviewed (approved):
I approve!
